Add deposit for setting session keys

Krayt78 opened the pull request:
# Description
Implement reserved balance tracking for test accounts in the session pallet. 
Update KeyDeposit type to use BalanceOf. Add tests for setting and purging keys to verify fund reservation and unreservation.
Update inner_set_keys to accept old keys as an argument to reduce redundant storage reads and improve efficiency in the session pallet.

xlc reviewed (commented):
we really need to deprecate unnamed reserves. should use names reserve instead

Krayt78 commented:
> we really need to deprecate unnamed reserves. should use names reserve instead

i wil l change to use named reserve instead then, but from what i can see i still need to use ReservableCurrency for the can_reserve function since NamedReservableCurrency doesnt seem to implement it

bkchr commented on substrate/frame/session/src/lib.rs:
```diff
@@ -436,8 +441,21 @@ pub mod pallet {
 		type WeightInfo: WeightInfo;
+		
+		/// The currency type for reserving when setting keys.
+		type Currency: NamedReservableCurrency<Self::AccountId>;
+		
+		/// The reserve identifier type.
+		type ReserveIdentifier: Get<<Self::Currency as NamedReservableCurrency<Self::AccountId>>::ReserveIdentifier> + TypeInfo + 'static;
```
We should directly use [`Holds`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/tokens/fungibles/hold/trait.Mutate.html#method.hold)

Krayt78 replied:
i have just pushed a fix to replace Reserve by Hold instead

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -558,6 +576,10 @@ pub mod pallet {
 		ValidatorDisabled { validator: T::ValidatorId },
 		/// Validator has been re-enabled.
 		ValidatorReenabled { validator: T::ValidatorId },
+		/// Funds have been held for setting keys.
```
No need for these -- balances already emits events.

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -573,6 +595,8 @@ pub mod pallet {
 		/// Key setting account is not live, so it's impossible to associate keys.
 		NoAccount,
+		/// Insufficient funds for setting keys.
+		InsufficientFunds,
```
Same.

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -836,10 +860,39 @@ impl<T: Config> Pallet<T> {
+		if is_new_registration {
+			let deposit = T::KeyDeposit::get();
+			let reason = T::HoldReason::get();
+			ensure!(
```
If this is called from within a transaction, which AFAIK it is, no need to do this check first -- you can directly call `T::Currency::hold(&reason, account, deposit)?;`. If it fails, everything is rolled back. 

Recall that FRAME is transactional ;)

Krayt78 replied:
Thanks :D

kianenigma commented on substrate/frame/session/src/mock.rs:
```diff
@@ -267,10 +290,303 @@ impl Config for Test {
 	type FullIdentificationOf = sp_runtime::traits::ConvertInto;
 }
+
+pub mod pallet_balances {
```
This is all very complicated -- You can just add pallet-balances here.

Krayt78 commented:
@serban300 @Ank4n Hi guys ! I have added pallet balances to the beefy mmr due to it being present in pallet sessions now and the test extract_validation_context_should_work_correctly now breaks. Its a hardcoded string, how did you guys make it in the first place ? just to make sure that i didnt break something and its just that the byte array is different because i added to the frame system. 

Here is the part that breaks : 
https://github.com/paritytech/polkadot-sdk/blame/c679ed79a1de575d4c3d5b8878b3e401b024f2f2/substrate/frame/beefy-mmr/src/tests.rs#L216-L222

Thanks in advance !

Krayt78 commented:
> > > @serban300 @Ank4n Hi guys ! I have added pallet balances to the beefy mmr due to it being present in pallet sessions now and the test extract_validation_context_should_work_correctly now breaks. Its a hardcoded string, how did you guys make it in the first place ? just to make sure that i didnt break something and its just that the byte array is different because i added to the frame system.
> > > Here is the part that breaks : https://github.com/paritytech/polkadot-sdk/blame/c679ed79a1de575d4c3d5b8878b3e401b024f2f2/substrate/frame/beefy-mmr/src/tests.rs#L216-L222
> > > Thanks in advance !
> > 
> > 
> > Probably @serban300 knows this better, but this is expected if something changes in your block (such as a new event being emitted).
> 
> Yes, yes this is expected if you added a new event or something to the block or block header

Yeah thats what i figured, i am wondering tho if i should just blindly change the bytes value to the new one or if there is somethign else to test ? (i think its fine as is and i could just swap the old bytes for the new one)

serban300 commented:
> Yeah thats what i figured, i am wondering tho if i should just blindly change the bytes value to the new one or if there is somethign else to test ? (i think its fine as is and i could just swap the old bytes for the new one)

Yes, you can just change the old bytes to the new ones

kianenigma commented on cumulus/pallets/collator-selection/src/mock.rs:
```diff
@@ -140,6 +140,9 @@ impl pallet_session::Config for Test {
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU64<0>;
```
```suggestion
	type KeyDeposit = ();
```

kianenigma commented on cumulus/parachains/runtimes/assets/asset-hub-rococo/src/lib.rs:
```diff
@@ -836,6 +836,9 @@ impl pallet_session::Config for Runtime {
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU128<0>;
```
```suggestion
	type KeyDeposit = ();
```

kianenigma commented on cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs:
```diff
@@ -890,6 +890,9 @@ impl pallet_session::Config for Runtime {
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU128<0>;
```
```suggestion
	type KeyDeposit = ();
```

kianenigma commented on cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:
```diff
@@ -506,6 +506,9 @@ impl pallet_session::Config for Runtime {
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU128<0>;
```
```suggestion
	type KeyDeposit = ();
```

kianenigma commented on cumulus/parachains/runtimes/bridge-hubs/bridge-hub-westend/src/lib.rs:
```diff
@@ -492,6 +492,9 @@ impl pallet_session::Config for Runtime {
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU128<0>;
```
```suggestion
	type KeyDeposit = ();
```

kianenigma replied:
and the rest please.

kianenigma commented on substrate/bin/node/runtime/src/lib.rs:
```diff
@@ -688,8 +689,10 @@ impl pallet_session::Config for Runtime {
 	type WeightInfo = pallet_session::weights::SubstrateWeight<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
+	type KeyDeposit = ConstU128<0>;
```
```suggestion
	type KeyDeposit = ConstU128<UNIT>;
```

This can actually be non-zero

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -400,6 +401,10 @@ pub mod pallet {
 
+	/// A simple identifier for session keys hold reason.
+	#[derive(codec::Encode, codec::Decode, codec::MaxEncodedLen, Debug, PartialEq, Eq, Clone)]
+	pub struct SessionKeysHoldReason;
```
```suggestion
	pub enum SessionHoldReason {
	    #[codec(index = 0)]
	    Keys
};
```

This will be forward compatible in case we add more "reasons".

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -400,6 +401,10 @@ pub mod pallet {
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
+	/// A simple identifier for session keys hold reason.
+	#[derive(codec::Encode, codec::Decode, codec::MaxEncodedLen, Debug, PartialEq, Eq, Clone)]
+	pub struct SessionKeysHoldReason;
+
```
```suggestion
```

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -516,6 +536,14 @@ pub mod pallet {
+	pub enum HoldReason {
+		// Funds are held when settings keys
+		#[codec(index = 0)]
+		SettingKeys,
```
```suggestion
		Keys,
```

kianenigma commented on substrate/frame/session/src/mock.rs:
```diff
@@ -275,3 +294,21 @@ impl crate::historical::Config for Test {
 	type FullIdentificationOf = sp_runtime::traits::ConvertInto;
 }
+
+#[derive_impl(pallet_balances::config_preludes::TestDefaultConfig as pallet_balances::DefaultConfig)]
```
Please remove the stuff that is already derived from this line?

Krayt78 replied:
thx ! I was just mimicking the other pallets and didnt know what it was for. Now i know :)

kianenigma commented on substrate/frame/session/src/tests.rs:
```diff
@@ -467,6 +467,91 @@ fn test_migration_v1() {
+		// Account 999 is mocked to have ExistentialDeposit but insufficient funds
+		let account_id = 999;
+		let keys = MockSessionKeys { dummy: UintAuthorityId(account_id).into() };
+		frame_system::Pallet::<Test>::inc_providers(&account_id);
```
Better is to fund 999, but lower than the set KeyDeposit.

kianenigma commented on substrate/frame/session/src/tests.rs:
```diff
@@ -467,6 +467,91 @@ fn test_migration_v1() {
+		let keys = MockSessionKeys { dummy: UintAuthorityId(account_id).into() };
+		frame_system::Pallet::<Test>::inc_providers(&account_id);
+		// Make sure we have a validator ID
+		ValidatorAccounts::mutate(|m| {
```
@Ank4n another follow-up here would be to also remove this, agreed? The author of this PR can do it as a next one :)

kianenigma commented on substrate/frame/session/src/tests.rs:
```diff
@@ -467,6 +467,91 @@ fn test_migration_v1() {
+		});
+
+		// Attempt to set keys with an account that has insufficient funds
+		let res = Session::set_keys(RuntimeOrigin::signed(account_id), keys, vec![]);
```
```suggestion
		assert_err!(Session::set_keys(RuntimeOrigin::signed(account_id), keys, vec![], sp_runtime::TokenError::FundsUnavailable), );
```

kianenigma reviewed (commented):
Logic is def good, but some cleanup is needed.

Krayt78 commented:
for some reason when the CI runs the taplo format he has this :
```
INFO taplo:format_files:collect_files: found files total=631 excluded=46 cwd="/__w/polkadot-sdk/polkadot-sdk"
ERROR taplo:format_files: the file is not properly formatted path="/__w/polkadot-sdk/polkadot-sdk/cumulus/parachains/runtimes/testing/yet-another-parachain/Cargo.toml"
```

When i do it locally i have this :
` INFO taplo:format_files:collect_files: found files total=643 excluded=58 `
 
 So idk why i am expluding more files than the CI since my branch is up to date with master and i am using the right command `taplo format --check --config .config/taplo.toml`

kianenigma reviewed (approved):
Did some final cleanups. LGTM

re-gius reviewed (approved):
LGTM

Krayt78 commented on prdoc/pr_7953.prdoc:
```diff
@@ -0,0 +1,65 @@
+title: Add deposit for setting session keys
+doc:
+- audience: Runtime Dev
+  description: |-
+    Implement hold balance tracking for test accounts in the session pallet.
```
like so ? @bkchr 

```suggestion
- audience: [Runtime Dev, Runtime User]
  description: |-
    Implement hold balance tracking for test accounts in the session pallet.
    Setting session keys will put some funds on hold in the future.
```

Krayt78 commented:
> @Krayt78 last attempts to make the CI green please :)
> 
> Please put your polkadot address in the PR description for a tip.

yup ! will ping you once its good, those cargo clippys take forever on my computer :')

Krayt78 commented on cumulus/pallets/collator-selection/src/mock.rs:
```diff
@@ -140,6 +140,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on cumulus/parachains/runtimes/assets/asset-hub-rococo/src/lib.rs:
```diff
@@ -837,6 +837,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:
```diff
@@ -507,6 +507,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on cumulus/parachains/runtimes/bridge-hubs/bridge-hub-westend/src/lib.rs:
```diff
@@ -493,6 +493,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on cumulus/parachains/runtimes/collectives/collectives-westend/src/lib.rs:
```diff
@@ -506,6 +506,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on cumulus/parachains/runtimes/people/people-rococo/src/lib.rs:
```diff
@@ -368,6 +368,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on polkadot/runtime/rococo/src/lib.rs:
```diff
@@ -484,6 +484,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on polkadot/runtime/westend/src/lib.rs:
```diff
@@ -538,6 +538,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = pallet_session::disabling::UpToLimitWithReEnablingDisablingStrategy;
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/authority-discovery/src/lib.rs:
```diff
@@ -212,6 +213,9 @@ mod tests {
 		type DisablingStrategy = ();
 		type WeightInfo = ();
+		type Currency = Balances;
+		type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/babe/src/mock.rs:
```diff
@@ -102,6 +102,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/beefy/src/mock.rs:
```diff
@@ -186,6 +186,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/election-provider-multi-phase/test-staking-e2e/src/mock.rs:
```diff
@@ -145,6 +145,9 @@ impl pallet_session::Config for Runtime {
 	>;
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/grandpa/src/mock.rs:
```diff
@@ -106,6 +106,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/im-online/src/mock.rs:
```diff
@@ -129,6 +136,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/offences/benchmarking/src/mock.rs:
```diff
@@ -97,6 +97,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/root-offences/src/mock.rs:
```diff
@@ -168,6 +168,9 @@ impl pallet_session::Config for Test {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/staking-async/ahm-test/src/rc/mock.rs:
```diff
@@ -179,6 +179,9 @@ impl pallet_session::Config for Runtime {
 
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/staking-async/runtimes/parachain/src/lib.rs:
```diff
@@ -908,6 +908,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = weights::pallet_session::WeightInfo<Runtime>;
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented on substrate/frame/staking/src/mock.rs:
```diff
@@ -151,6 +151,9 @@ impl pallet_session::Config for Test {
 		pallet_session::disabling::UpToLimitWithReEnablingDisablingStrategy<DISABLING_LIMIT_FACTOR>;
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 replied:
```suggestion
```

Krayt78 commented on templates/parachain/runtime/src/configs/mod.rs:
```diff
@@ -277,6 +277,9 @@ impl pallet_session::Config for Runtime {
 	type DisablingStrategy = ();
 	type WeightInfo = ();
+	type Currency = Balances;
+	type RuntimeHoldReason = RuntimeHoldReason;
```
```suggestion
```

Krayt78 commented:
@kianenigma its good besides CI complaining that PRdoc should be Major instead of Patch

kianenigma commented on substrate/frame/session/src/lib.rs:
```diff
@@ -440,6 +445,19 @@ pub mod pallet {
+
+		/// The currency type for placing holds when setting keys.
+		type Currency: Inspect<Self::AccountId>
+			+ HoldMutate<Self::AccountId, Reason = Self::RuntimeHoldReason>;
```
Nice, is this something new https://blog.rust-lang.org/2024/06/13/Rust-1.79.0/#bounds-in-associated-type-position

We can get rid of a bunch things like 

```
type RuntimeCall: ..
type CurrencyBalance: .. 
```

And so on now.

kianenigma commented:
/tip medium

mordamax commented:
/tip medium
